Add unknown property warning for use of `autofocus`

hkal opened the pull request:
Resolves #3248

This PR also:
  * Resurrects property warnings when `ReactDOMFeatureFlags.useCreateElement` is on
  * Adds property validity event to debug tool

I wrote some tests for this work, but didn't notice any other `__DEV__` specific warnings being tested. I can include them with this PR if they're wanted.

hkal commented:
@jimfb - noticed set the label on the original bug. Let me know if this adequately addresses the problem.

gaearon commented:
Hmm. I think the plan was to move these warnings to element creation: #6345
cc @jimfb

jimfb commented:
@hkal Sorry for letting this one slip through the cracks.  Can you rebase and move it to element creation like in #6345 and then I think this is ready to merge.

hkal commented:
@jimfb - No worries.

Just so I'm clear, is the direction to consolidate `onCreateMarkupForProperty`, `onSetValueForProperty`, `onDeleteValueForProperty` and the event I added into a single event that is fired on element creation?

cc @gaearon

gaearon commented:
I think so. Afaik this event already exists, check out #6800. You might want to tweak that code.

hkal commented:
@gaearon You're right, looks like the event is already there. I removed the unnecessary code and updated the tests.

@jimfb Am I missing anything, or is this good?

Aweary commented:
```js
getPossibleStandardName: __DEV__ ? {autofocus: 'autoFocus'} : null,
```

@hkal `getPossibleStandardName` is [injected in `DOMProperty`](https://github.com/facebook/react/blob/416b5ef17343acebfd19e1436e3b2990e99e1d0e/src/renderers/dom/shared/DOMProperty.js#L112-L118) using [`HTMLDOMPropertyConfig.DOMAttributeNames`](https://github.com/facebook/react/blob/cccef3c68310df3bd611df2a7b98a530645c67c0/src/renderers/dom/shared/HTMLDOMPropertyConfig.js#L204-L209). Could you move this into that object instead so it's injected like the other values are? I think this can be merged after that change, as far as I can see.

Aweary commented:
@hkal ping! Would you by chance have any free time soon to make this small change? I'd like to get this in soon. If not, no worries, I can get https://github.com/facebook/react/pull/7471 in 👍

hkal commented:
@Aweary my bad. Somehow I missed these notifications. Updated the pull request, let me know if anything else needs changing.

Aweary commented:
Thanks @hkal, looks good to me 👍

zpao commented:
I'm actually going to revert this - it actually has more impact than just the warning and changes how autoFocus handling will behave. I can't demonstrate it in jsfiddle because frames and focus don't play nicely. But just take an example and render 2 inputs with autoFocus={true}. 15.x will consistently focus the 2nd. After this change the first will get focused (in Chrome) and we're otherwise subject to whatever the browser decides to do when nodes are updated or removed. That's not to say we should necessarily be handling this ourselves, but for the time being we shouldn't change it.

When we get to removing the whitelist we'll have to deal with that side of things. For now, if we want this to work, we need to handle this without adding the property to the whitelist as we don't want it to end up in the DOM.

Aweary commented:
Thanks for clarifying @zpao, I had tested that autoFocus still worked before merging but didn't notice the order may have been inconsistent.

hkal commented:
Thanks for explaining the issue, @zpao. Didn't even think to test that case.

@Aweary I'm pretty confident we can still get the warning without any unintentional side affects if we go back to pre-defining the autofocus property in `getPossibleStandardName`. Aside from attribute injection, the only call to it is for [producing warnings](https://github.com/facebook/react/blob/178cb7d3392505cd393b68e4e061c6ef860344a1/src/renderers/dom/shared/hooks/ReactDOMUnknownPropertyHook.js#L57).

Aweary commented:
@zpao are there any other properties like this that we're unable to validate right now because of how we inject values used in `validateProperty` based on a whitelist?

If not @hkal I think providing `autoFocus` as a default in `getPossibleStandardName` is a decent solution for now. IMO ideally validating a property would be independent of any whitelist, but if there aren't many other cases like this it's probably not worth addressing now since the whitelist is on its way out.
